### ISSUE TITLE
Enforcement of units and basis context usage

### ISSUE THREAD

**Issue by tmancal74** [tmancal74/quantarhei#124]
It is desirable not to allow functions that set physical values to be used outside the units context (energy_units etc.). These functions should raise an Exception if used without the context. Similarly, some functions should not be used inside the eigenbasis_of context.  Again, an Exception should be raised if this context is applied on the function.

**Comment by tmancal74** [comment]
The most elegant solution would be to implement decorators which decorate the functions for which the presence of context should be enforced (or prevented). The decorated function will detect the present of the context and through an exception if the context policy is breached.

**Comment by tmancal74** [comment]
Simple detection of eigenbasis_of context is implemented. Wrappers prevent_basis_context and enforce_basis_context are available and working. 

The new wrapper is not yet applied to anything as its presence will likely break several existing automated tests.

**Comment by detrin** [comment]
@tmancal74 I looked at this together with #444 and wanted to propose an alternative before more decorators get added, because I think there is a way to get full coverage with less code.

All links below point to current master ([`0b7d6d9`](https://github.com/tmancal74/quantarhei/commit/0b7d6d98b1f58f77505133595d2d7de4b10a284e)).

### How it works today

Every energy a user passes in ends up in [`Manager.convert_energy_2_internal_u`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/core/managers.py#L618-L634) (and its siblings for [frequency](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/core/managers.py#L654) and [length](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/core/managers.py#L680)), either directly or through [`UnitsManaged`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/core/managers.py#L806) / [`_TypedUnitsManaged.convert_2_internal_u`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/core/managers.py#L837). Outside a context the current unit is `int`, so the value is taken as rad/fs without any signal.

The only guard is per class: [`enforce_energy_units_context`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/core/wrappers.py#L82) is applied to [`CorrelationFunction`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/qm/corfunctions/correlationfunctions.py#L114) and [`LineshapeFunction`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/qm/corfunctions/correlationfunctions.py#L815), gated by [`_enforce_contexts`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/core/managers.py#L230). Every other class that accepts energies is unguarded.

### API today vs proposed

```python
qr.Molecule([0.0, 12000.0])          # today: silently 12000 rad/fs
                                     # proposed: DeprecationWarning now, error later

with qr.energy_units("1/cm"):        # unchanged, the normal way
    m = qr.Molecule([0.0, 12000.0])

with qr.energy_units("int"):         # unchanged, explicit internal units
    m = qr.Molecule([0.0, 2.26])

qr.Manager().set_current_units("energy", "1/cm")   # unchanged, counts as explicit
m = qr.Molecule([0.0, 12000.0])
```

Nothing changes for code that already states its units. Reading values back (`*_2_current_u`) is not checked.

### What changes internally

1. One check in the Manager, called at the top of each `convert_*_2_internal_u`:

```python
def _require_explicit_units(self):
    if self._units_policy == "off" or self._in_eu_count or self._units_set_globally:
        return
    if self._units_policy == "error":
        raise QuantarheiError("Energy passed without units; use energy_units(...)")
    warnings.warn("Energy passed without units ...", DeprecationWarning, stacklevel=_user_frame())

def convert_energy_2_internal_u(self, val):
    self._require_explicit_units()
    ...
```

`_units_policy` (`"off" | "warn" | "error"`) replaces the boolean `_enforce_contexts`. `_user_frame()` is a small helper that walks out of the `quantarhei` package so the warning points at the user's line (`skip_file_prefixes` would do this, but needs Python 3.12).

2. `enforce_energy_units_context` and `prevent_energy_units_context` are deleted, since the check above already covers every class. The basis decorators ([`prevent_basis_context`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/core/wrappers.py#L41), [`enforce_basis_context`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/core/wrappers.py#L55)) stay, because basis handling has no single entry point.

3. One rule inside the library: convert only at the user boundary, and never feed a value that is already internal back into `convert_*_2_internal_u`. For example, [`abscalculator.py:179`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/spectroscopy/abscalculator.py#L179) converts the RWA energies to current units in [`get_RWA_skeleton`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/qm/hilbertspace/hamiltonian.py#L148-L154) and then back again:

```python
# today
HR = HH.get_RWA_skeleton()
self.rwa = self.convert_2_internal_u(HR[Ne] - HR[Ng])
# proposed
self.rwa = HH.rwa_energies[Ne] - HH.rwa_energies[Ng]
```

This one happens to round-trip correctly, but the same pattern produces the bugs below.

4. Tests and doctests that pass internal values wrap their setup in `with qr.energy_units("int"):`.

### What it would touch

I instrumented the conversion functions and ran the unit tests and doctests from CI:

- 157 of 323 unit tests (38 files) and 69 of 120 doctests (22 modules) convert energies outside any units context. Most are fixed by one `with qr.energy_units("int"):` around the setup.
- 31 call sites in the library fire. About 25 are real user input (`Molecule`, `Mode`, `set_resonance_coupling`, `LabSetup`, ...), which is exactly where the check should be.
- About 5 are internal conversions of values the user never passed, which need small fixes: [`set_coupling_by_dipole_dipole`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/builders/aggregate_base.py#L305), the RWA round trip above, [`SpectralDensity.__add__`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/qm/corfunctions/spectraldensities.py#L444), and the duplicated conversion in [`Molecule.__init__`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/builders/molecules.py#L180) / [`OpenSystem.__init__`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/builders/opensystem.py#L74).

### Existing bugs this surfaced

All of them only show up inside a units context, so the documented usage in [`demo_050_PDB_FMO1.py`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/examples/demo_050_PDB_FMO1.py#L32-L37) is fine today. With FMO (`3eoj`, `BacterioChlorophyll`, expected 12500 cm⁻¹):

```python
model = BacterioChlorophyll(model_type="PDB")
with qr.energy_units("1/cm"):
    molecules = PDBFile("data_050_3eoj.pdb").get_Molecules(model=model)
```

gives 2.35 cm⁻¹ instead of 12500.

1. [`pdb.py:141`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/builders/pdb.py#L141) passes `model.default_energies`, already internal, to `Molecule(elenergies=...)`, which converts them again using the current units. This affects every model.
2. [`ChlorophyllA`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/models/chlorophylls.py#L27) / [`ChlorophyllB`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/models/chlorophylls.py#L142) pass `15200.0 * cm2int` (resp. 15700) through [`set_default_energies`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/models/molecularmodel.py#L26-L29), which converts again as well, so creating the model inside a `1/cm` context gives `default_energies[1] = 0.000539` instead of `2.863`. [`BacterioChlorophyll`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/models/bacteriochlorophylls.py#L21) sets the array directly and is not affected.
3. [`SpectralDensity.__add__`](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/qm/corfunctions/spectraldensities.py#L444) rebuilds the sum from `self.params`, which are already internal, and [converts them again](https://github.com/tmancal74/quantarhei/blob/0b7d6d98b1f58f77505133595d2d7de4b10a284e/quantarhei/qm/corfunctions/spectraldensities.py#L170). Adding two overdamped densities (20 and 40 cm⁻¹) inside `energy_units("1/cm")` gives `0.00624` at 100 cm⁻¹ instead of `0.00936`, the same value you get outside a context.

This matters for the proposal: once enforcement is on, `get_Molecules` outside a context would warn, and the natural reaction of wrapping it in `energy_units("1/cm")` is exactly the broken case. So these need to be fixed first. I can send a small separate PR for them regardless of what we decide here.

### Proposed order

1. Fix the three bugs above (independent).
2. Check in `"warn"` mode plus the ~5 internal fixes.
3. Migrate tests and doctests, package by package, then fail CI on the warning.
4. Default to `"error"` in a later release.

Would this direction work for you, or do you prefer to keep enforcement explicit per class as in #444? Happy to adapt either way.
